File: framework/languages/node/e2e/ResilienceLifecycle/Client/Scenarios/rl-d3-dispatch-error-evidence-scenario.ts
// RL-D3: Public logging sink에서 dispatch error를 확인한다 시나리오를 검증한다.
import type { RequestFailureRes } from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { postJson } from '../../../http-client';
import { waitForProviderEvidenceLine } from '../Support/provider-evidence';
import { profileReq, waitForAnyProviderTraffic } from '../Support/resilience-helpers';
import { ensure } from '../Support/scenario-assert';

export async function runRlD3(options: ClientOptions): Promise<void> {
  await waitForAnyProviderTraffic(options.consumerUrl, 'rl-d3-ready');
  const failed = await postJson<RequestFailureRes>(
    options.consumerUrl,
    '/profile/missing-request',
    profileReq('rl-d3-missing')
  );
  ensure(failed.failed, 'RL-D3 expected missing request handler failure.');
  ensure(
    failed.failureMessage.includes('request handler is registered'),
    `RL-D3 request did not reach the remote no-handler path: ${failed.failureMessage}`
  );

  const line = await waitForProviderEvidenceLine(
    options,
    (entry) => entry.includes('dispatch-error|') && entry.includes('packet_name=MissingProfileReq'),
    'RL-D3 dispatch-error evidence timed out.'
  );
  ensure(
    line.includes('outcome=failed')
      && line.includes('reason=no_handler')
      && line.includes('action=reply_error')
      && line.includes('packet_name=MissingProfileReq'),
    'RL-D3 dispatch-error evidence did not preserve the common failure fields.'
  );

  console.log('scenario RL-D3 passed');
}

File: framework/languages/dotnet/e2e/ResilienceLifecycle/Client/Scenarios/RlD3DispatchErrorEvidenceScenario.cs
// Verifies RL-D3 Dispatch Error Evidence behavior.
using ResilienceLifecycle.Client.Support;
using ResilienceLifecycle.Shared;
using Zlink.HttpClient;

namespace ResilienceLifecycle.Client.Scenarios;

// RL-D3 verifies dispatch-error evidence for missing send handling.
internal static class RlD3DispatchErrorEvidenceScenario
{
    public static async Task RunAsync(
        ZLinkHttpClient consumer,
        ZLinkHttpClient providerA,
        ZLinkHttpClient providerB)
    {
        var failed = await consumer.Post("/profile/request/missing")
            .Body(new ProfileReq("fast", "rl-d3-missing"))
            .AsyncRaw();
        ZlinkStreamAssert.Ensure(failed.Status >= 500, "RL-D3 expected missing request handler failure.");

        {
            using var timeout = new CancellationTokenSource(TimeSpan.FromSeconds(15));
            var request = new EvidenceWaitReq(["dispatch-error|"], [["packet=MissingProfileReq"]]);
            var waitA = providerA.Post("/evidence/wait").Body(request).Async<string[]>(timeout.Token).AsTask();
            var waitB = providerB.Post("/evidence/wait").Body(request).Async<string[]>(timeout.Token).AsTask();
            var completed = await Task.WhenAny(waitA, waitB);
            var evidence = (await completed).Body;
            timeout.Cancel();
            ZlinkStreamAssert.Ensure(
                evidence.Any(line => line.Contains("packet=MissingProfileReq", StringComparison.Ordinal)),
                "RL-D3 dispatch-error evidence did not include MissingProfileReq.");
        }

        Console.WriteLine("scenario RL-D3 passed");
    }
}